editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -0,0 +1,191 @@
+        return
+      }
+    })
+
+    const fixedTopLevelElements = applyUtopiaJSXComponentsChanges(
+      newParsed.topLevelElements,
+      workingComponents,
+    )
+
+    const fixedHighlightBounds = mapArrayToDictionary(
+      Object.values(newParsed.highlightBounds),
+      (highlightBound) => {
+        const uidFix = newToOldUidMapping[highlightBound.uid]
+        if (uidFix != null) {
+          return uidFix.oldUID
+        } else {
+          return highlightBound.uid
+        }
+      },
+      (highlightBound) => {
+        const uidFix = newToOldUidMapping[highlightBound.uid]
+        if (uidFix != null) {
+          return {
+            ...highlightBound,
+            uid: uidFix.oldUID,
+          }
+        } else {
+          return highlightBound
+        }
+      },
+    )
+
+    return {
+      ...newParsed,
+      topLevelElements: fixedTopLevelElements,
+      highlightBounds: fixedHighlightBounds,
+    }
+  }
+}
+
+function zipTopLevelElements(
+  firstTopLevelElements: Array<TopLevelElement>,
+  secondTopLevelElements: Array<TopLevelElement>,
+  onElement: (
+    firstContents: JSXElementChild | null,
+    secondContents: JSXElementChild,
+    newTemplatePath: StaticElementPath,
+  ) => void,
+): void {
+  secondTopLevelElements.forEach((newTopLevelElement, index) => {
+    if (isUtopiaJSXComponent(newTopLevelElement)) {
+      const uid = getUtopiaID(newTopLevelElement.rootElement)
+
+      const oldTopLevelElement = firstTopLevelElements.find((tle): tle is UtopiaJSXComponent => {
+        return isUtopiaJSXComponent(tle) && getUtopiaID(tle.rootElement) === uid
+      })
+
+      if (oldTopLevelElement != null) {
+        walkElementChildren(
+          TP.emptyElementPath,
+          [oldTopLevelElement.rootElement],
+          [newTopLevelElement.rootElement],
+          onElement,
+        )
+      }
+    }
+  })
+}
+
+function walkElementChildren(
+  pathSoFar: StaticElementPath,
+  oldElements: Array<JSXElementChild>,
+  newElements: Array<JSXElementChild>,
+  onElement: (
+    firstContents: JSXElementChild | null,
+    secondContents: JSXElementChild,
+    newTemplatePath: StaticElementPath,
+  ) => void,
+): void {
+  /**
+   * this first version works by trying to match up indexes. this is really primitive.
+   * here's some ideas how could we improve it
+   * • before correcting a UID, make sure the uids of the preceding and proceding siblings are matching! this makes the comparison more robust and makes us less exposed to off-by-one errors
+   * • try to match offsets: if the user deletes or inserts an element, all subsequent uids will be shifted and thus mismatched in an index-to-index comparison
+   */
+
+  newElements.forEach((newElement, index) => {
+    const oldElement: JSXElementChild | null = oldElements[index]
+    const newUid = getUtopiaID(newElement)
+    const path = TP.appendToElementPath(pathSoFar, newUid)
+
+    onElement(oldElement, newElement, path)
```

Comment:
I think this call should be inside the below `if` statement (or at least should be wrapped in a conditional checking that the new and old elements are `JSXElement`s

Reply:
fixed